### PR TITLE
[Intel MKL] Adding support for AvgPool with native format

### PR DESCRIPTION
**Pull request by mahmoud-abuzaina**

### CODE REVIEW

File: tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc [hunk 1]
```diff
@@ -54,8 +54,13 @@ void MklPoolingFwdPrimitive<T>::Setup(const MklPoolingParams& fwdParams) {
...
 #endif  //  !ENABLE_MKLDNN_V1
-  context_.dst_md.reset(new memory::desc({fwdParams.dst_dims}, MklDnnType<T>(),
-                                         MEMORY_FORMAT::any));
+  if (!fwdParams.native_format) {
+    context_.dst_md.reset(new memory::desc(
+        {fwdParams.dst_dims}, MklDnnType<T>(), MEMORY_FORMAT::any));
+  } else {
+    context_.dst_md.reset(new memory::desc(
+        {fwdParams.dst_dims}, MklDnnType<T>(), fwdParams.src_format));
+  }
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc:63]: Nit: I think this is easier to read:
```suggestion
  context_.dst_md.reset(new memory::desc(
      {fwdParams.dst_dims}, MklDnnType<T>(),
      fwdParams.native_format ? fwdParams.src_format : MEMORY_FORMAT::any));
```

File: tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc [hunk 2]
```diff
@@ -187,8 +192,13 @@ void MklPoolingBwdPrimitive<T>::Setup(const MklPoolingParams& bwdParams) {
...
   context_.src_md.reset(new memory::desc(bwdParams.src_md.data));
-  context_.dst_md.reset(new memory::desc({bwdParams.dst_dims}, MklDnnType<T>(),
-                                         MEMORY_FORMAT::any));
+  if (!bwdParams.native_format) {
+    context_.dst_md.reset(new memory::desc(
+        {bwdParams.dst_dims}, MklDnnType<T>(), MEMORY_FORMAT::any));
+  } else {
+    context_.dst_md.reset(new memory::desc(
+        {bwdParams.dst_dims}, MklDnnType<T>(), bwdParams.src_format));
+  }
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc:201]: Nit: Same here.
```suggestion
  context_.dst_md.reset(new memory::desc(
      {bwdParams.dst_dims}, MklDnnType<T>(),
      bwdParams.native_format ? bwdParams.src_format : MEMORY_FORMAT::any));
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 3]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Input("value: T")
+    .Output("output: T")
+    .Attr("ksize: list(int) >= 4")
+    .Attr("strides: list(int) >= 4")
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr("T: {float, half, double, bfloat16}")
+    .SetShapeFn(shape_inference::AvgPoolShape)
+    .Doc(R"doc(
+MKL version of AvgPool operator that does not depend
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:182]: Nit:
```suggestion
oneDNN version of AvgPool operator that does not depend
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 4]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnetDataFormatAttrString())
+    .Attr("T: {float, half, double, bfloat16}")
+    .SetShapeFn([](InferenceContext* c) {
+      ShapeHandle s;
+      TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(0, &s));
+      TF_RETURN_IF_ERROR(c->WithRank(s, 4, &s));
+      c->set_output(0, s);
+      return Status::OK();
+    })
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:205]: This is the same shape function as `_MklAvgPoolGrad` and `AvgPoolGrad`. Please refactor and put the function in `//tensorflow/core/framework/common_shape_fns.cc`.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 5]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Doc(R"doc(
+MKL version of AvgPoolGrad operator that does not depend
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:207]: Nit:
```suggestion
oneDNN version of AvgPoolGrad operator that does not depend
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 6]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Input("value: T")
+    .Output("output: T")
+    .Attr("ksize: list(int) >= 5")
+    .Attr("strides: list(int) >= 5")
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnet3dDataFormatAttrString())
+    .Attr("T: {float, half, double, bfloat16}")
+    .SetShapeFn(shape_inference::Pool3DShape)
+    .Doc(R"doc(
+MKL version of AvgPool3D operator that does not depend
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:225]: Nit:
```suggestion
oneDNN version of AvgPool3D operator that does not depend
```

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 7]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Attr(GetPaddingAttrString())
+    .Attr(GetConvnet3dDataFormatAttrString())
+    .Attr("T: {float, half, double, bfloat16}")
+    .SetShapeFn([](InferenceContext* c) {
+      ShapeHandle s;
+      TF_RETURN_IF_ERROR(c->MakeShapeFromShapeTensor(0, &s));
+      TF_RETURN_IF_ERROR(c->WithRank(s, 5, &s));
+      c->set_output(0, s);
+      return Status::OK();
+    })
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:248]: This is the same as the shape functions of `_MklAvgPool3DGrad` and `AvgPool3DGrad`. Please refactor it and put the function in `//tensorflow/core/framework/common_shape_fns.cc`.

File: tensorflow/core/ops/mkl_nn_ops.cc [hunk 8]
```diff
@@ -169,6 +169,92 @@ REGISTER_OP("_MklPadWithFusedConv2D")
...
+    .Doc(R"doc(
+MKL version of AvgPool3DGrad operator that does not depend
```
Comment (penpornk) [inline tensorflow/core/ops/mkl_nn_ops.cc:250]: Nit:
```suggestion
oneDNN version of AvgPool3DGrad operator that does not depend
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the delay!

File: tensorflow/core/kernels/mkl/mkl_avgpooling_op.cc [hunk 9]
```diff
@@ -274,7 +282,8 @@ class MklAvgPoolingGradOp : public MklPoolingBackwardOpBase<T> {
           pooling_bwd->GetPoolingBwdPd();
       T* diff_dst_data = nullptr;
       if (IS_DIFF_DST_REORDER_NEEDED(diff_dst_md, pooling_bwd_pd,
-                                     pooling_bwd)) {
+                                     pooling_bwd) &&
+          !this->native_format_) {
```
[the remark this answers is not in the document]
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_avgpooling_op.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc (hunk 1 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc]: Done.

File: tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc (hunk 2 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/kernels/mkl/mkl_pooling_ops_common.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 3 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 4 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 5 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 6 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

File: tensorflow/core/ops/mkl_nn_ops.cc (hunk 8 above)
Comment (mahmoud-abuzaina) [reply tensorflow/core/ops/mkl_nn_ops.cc]: Done.

**Comment by mahmoud-abuzaina** [comment]
Thank you for reviewing the PR. I have addressed your comments.

**Review by penpornk — APPROVED** [review]
Thank you for the changes!

**Pull request merged**